src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c

```diff
@@ -585,13 +585,15 @@ AndroidCryptoNative_SSLStreamRead(SSLStream* sslStream, uint8_t* buffer, int32_t
 
     if (rem > 0)
     {
-        data = make_java_byte_array(env, rem);
+        int32_t bytes_to_read = rem < length ? rem : length;
+        assert(bytes_to_read == rem && "destination buffer is not large enough");
```

Comment:
Is this assert a debug-only assert, or a release-and-debug assert?

If it's debug only, what state would we be in if it hit in the real world?  I think it'd be data loss, since I think ByteBuffer.compact will throw away the excess data.

Reply:
This assert is debug-only. I don't believe the code would lose data even if it ran into this edge case because `compact` removes only the data that was already transferred into `data` and it would [move the remaining data to the beginning of the `appInBuffer`](https://developer.android.com/reference/java/nio/ByteBuffer#compact()).

From what I can tell, the function is currently used to read all of the data at once (the sizes of the input and output buffer are the same and it is assumed that the size of the output is never larger than the input) and it never checks if there are any bytes left in the input buffer after this function is used. These assumptions might not hold if somebody decides to reuse this function in the future though.

Reply:
The second option would be to avoid reading any data from the buffer and report `SSLStreamStatus_Error`. Maybe that is actually a better option and it also won't affect the current use on this function.

Reply:
>  removes only the data that was already transferred into data 

Ah, right, since the destination array is smaller it'll bound the read.  That's probably reasonable, then.

If we're handling the case, we can probably drop the assert and let it just run.

Reply:
I doublechecked the series of `ByteBuffer` operations and you can obtain all of the data by repeatedly calling `AndroidCryptoNative_SSLStreamRead`. The behavior is now the same in the other two PAL implementations (`AppleCrypto`, `CryptoNative`) and it matches the description in `pal_sslstream.h` (`"Unless data from a previous incomplete read is present, this will invoke the STREAM_READER callback."`). The "second option" would not be a good idea after all.

I think this PR is ready to be merged. We unfortunately don't have any unit tests for the PAL layer AFAIK.